### PR TITLE
Implement bytecode instrumentation at deploy time

### PR DESCRIPTION
**Pull request by CPerezz**
As stated in #115 and #116 we need a way to instrumentate the WASM
bytecode of the contracts deployed to be able to inject gas_metering
tools as well as removal of floating point operations among other
checks.

This PR adds the tooling and instrumentation linked to the
`NetworksState::deploy()` fn adding the following checks and
instrumentation for the bytecode:
- No floating point ops are used in the contract.
- Gas metering tools are injected correctly and work.
- Stack height metering tools so that we never pass the maximum allowed.
- Table size limits used in the wasm Modules are not bigger than what is
allowed.
- Validity check of all of the previous instrumentation.

Resolves: #115, #116.

This leaves as a new issue to address: #174

### CODE REVIEW

File: src/state.rs [hunk 1]
```diff
@@ -75,14 +75,28 @@ where
...
+    ) -> Result<ContractId, VMError<S>> {
+        // Before any instrumentation is applied, generate the contract id.
+        let id: ContractId =
+            S::Ident::from_bytes(contract.code.as_ref()).into();
+
+        // Instrumentalize the contract.
+        ContractInstrumenter::instrument(
+            &mut contract,
+            &crate::Schedule::default(),
+        )?;
```
Comment (ZER0) [inline src/state.rs:93]: Basically you're using the struct as a function.
You could just have:
```rust
instrument(&mut contract)?;
```
(Since you get the default `Schedule` from the crate you don't need even to pass it).

More, since you have a function that takes a mutable contract, it could be directly a *method* of a contract:
```rust
contract.instrument()?;
```
You could also return a *new* contract instead of mutating the previous one or returns `self` anyway so that you can have:
```rust
.insert(id.clone(), contract.instrument()?)
```

That said, if you instead want to keep the `ContractInstrumentalizer` as public builder API, then you should have *here* something like:
```rust

// you can decide if you want to pass the `Schedule` or instantiate the default internally
let instrumenter = ContractInstrumenter::new(&mut contract)
                                  .with_grow_cost() // this is a wrapper to `ruleset`, that you create in `new`
                                  .with_forbidden_floats() // same here
                                  .with_table_size_limit() // same here for the Schedule
                                  .with_gas_metering()
                                  .with_stack_metering();

instrumenter.inject()?.validate()?;
```

Basically the point is: or you encapsulate everything (and in that case you don't need the structure to be "known" from the outside world in terms of API), or you don't, and let the consumer the decision about how the instrumentalization should be done (with or without floats, with or without gas metering, etc).

File: tests/lib.rs [hunk 2]
```diff
@@ -93,6 +94,82 @@ fn counter_trivial() {
...
+
+    let code = include_bytes!(
+        "../target/wasm32-unknown-unknown/release/counter_float.wasm"
+    );
+
+    let contract = Contract::new(counter, code.to_vec(), &store).unwrap();
+
+    let mut network = NetworkState::<MS>::default();
+
+    assert!(network.deploy(contract).is_err());
```
Comment (ZER0) [inline tests/lib.rs:170]: I would actually check that this is the error you're expecting.
To do so, you can use a combination of `assert!` and `matches!` macros. See: https://doc.rust-lang.org/std/macro.matches.html

File: tests/lib.rs [hunk 3]
```diff
@@ -93,6 +94,82 @@ fn counter_trivial() {
...
+
+    let contract = Contract::new(counter, code.to_vec(), &store).unwrap();
+
+    let mut network = NetworkState::<MS>::default();
+
+    let contract_id = network.deploy(contract).expect("Deploy error");
+
+    let mut gas = GasMeter::with_limit(1);
+
+    assert!(network
```
Comment (ZER0) [inline tests/lib.rs:148]: I would actually check that this is the error you're expecting.
To do so, you can use a combination of `assert!` and `matches!` macros. See: https://doc.rust-lang.org/std/macro.matches.html

**Review by ZER0 — CHANGES_REQUESTED** [review]
LGTM, but there are few nits to address, and the public API to decide in which direction (encapsulate everything or not).

Since I'm not going to work on Monday and Tuesday, ask to @krl to do the next round of review (and his feedback would be appreciated anyway).

File: src/contract.rs [hunk 4]
```diff
@@ -48,3 +57,125 @@ impl Contract {
...
+    module: Module,
+    schedule: &'a crate::Schedule,
+    #[doc(hidden)]
+    _marker: PhantomData<S>,
+}
+
+impl<'a, S: Store> ContractInstrumenter<'a, S> {
+    /// Creates a new [`ContractInstrumenter`] instance from a mutable reference
+    /// to a [`Contract`] and a [`Schedule`] config.
+    pub(crate) fn instrument(
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply src/contract.rs:71]: Because I did not think at the moment that a trait would be a better solution and also on that way all of the instrumentalization code is self-contained in this struct impl block.

**Review by krl — APPROVED** [review]

**Comment by krl** [comment]
Looks good, have nothing to nitpick
